src/kernels/MIOpenBatchNormFwdInferPerAct.cl

```diff
@@ -35,53 +35,44 @@
-    const __global _FLOAT_PREC* __restrict scale,       /* gamma 1xCxHxW */
-    const __global _FLOAT_PREC* __restrict bias,        /* beta 1xCxHxW */
-    double epsilon)
+MIOpenBatchNormFwdInferPerActivationEst(const __global _FLOAT* in,
+                                        __global _FLOAT* __restrict out,
+                                        __global _FLOAT_PREC* __restrict estimatedMean,
+                                        __global _FLOAT_PREC* __restrict estimatedVariance,
+                                        const __global _FLOAT_PREC* __restrict scale,
+                                        const __global _FLOAT_PREC* __restrict bias,
+                                        double epsilon,
+                                        unsigned int batchSize,
+                                        unsigned int imageDims,
+                                        unsigned int batchStride)
 {
 
     // PER ACTIVATION
     _FLOAT_PREC mean, variance;
     _FLOAT_PREC invVariance, elemStd, inhat;
     _FLOAT_PREC pvt_scale, pvt_bias;
-    unsigned int adjIndex, inImgIndex, index;
-
-    int xgid    = get_global_id(0);
+    unsigned int adjIndex, index;
     int ygid    = get_global_id(1);
     int yglb_sz = get_global_size(1);
+    int grpid   = get_group_id(0);
 
-    int Cidx = MIO_BN_HW * xgid;
-
-    // move across the sections of an image in the mini_batch stack
-    for(int img_offset = 0; img_offset < MIO_BN_HW; img_offset += yglb_sz)
+    for(int img_offset = ygid; img_offset < imageDims; img_offset += yglb_sz)
     {
-        inImgIndex = img_offset + ygid;
-        if(inImgIndex < MIO_BN_HW)
-        {
-            adjIndex    = Cidx + inImgIndex; // gamma and beta tensor index
-            mean        = estimatedMean[adjIndex];
-            variance    = estimatedVariance[adjIndex];
-            invVariance = rsqrt(fabs(variance + epsilon));
-            pvt_scale   = *(scale + adjIndex);
-            pvt_bias    = *(bias + adjIndex);
+        adjIndex    = (grpid * imageDims) + img_offset;
+        mean        = estimatedMean[adjIndex];
+        variance    = estimatedVariance[adjIndex];
+        invVariance = rsqrt(fabs(variance + epsilon));
+        pvt_scale   = *(scale + adjIndex);
+        pvt_bias    = *(bias + adjIndex);
 
-#pragma unroll
-            for(int n = 0; n < MIO_BN_N; n++)
-            {
-                // per (x-dims) channel load a block of data into LDS
-                index   = MIO_BN_CHW * n + adjIndex;
-                elemStd = (_FLOAT_PREC)(*(in + index)) - mean; // (x_i - mean)
-                inhat   = elemStd * invVariance;
-                out[index] =
-                    (_FLOAT)(mad(pvt_scale, inhat, pvt_bias)); //	y_i = gamma*x_hat + beta
-            }                                                  // end for
-        }                                                      // end if
-    } // end for(img_offset) //image mini_batch is processed
```

Comment:
Does loop unroll improve any performance here?

Reply:
I can try, but there is not compiler hints, so it would be guessing.

Reply:
FYI Compiler team asked us to report all cases where manual loop unrolling gives some profit. The objective is to get rid of all manual unrolling.

Reply:
yes, We should report such cases.  We should keep loop unrolling if it improves performance for a range of configs..